atomic_reactor/config.py

```diff
@@ -0,0 +1,568 @@
+
+    def get_enabled_clusters_for_platform(self, platform):
+        cluster_configs = self._get_cluster_configuration()
+        if platform not in cluster_configs:
+            return []
+        return [conf for conf in cluster_configs[platform] if conf.enabled]
+
+    def cluster_defined_for_platform(self, platform):
+        cluster_configs = self._get_cluster_configuration()
+        return platform in cluster_configs
+
+    @property
+    def odcs_config(self):
+        """
+        Return an odcs config object created from the odcs config configured in
+        reactor config
+
+        :return: the object of ODCSConfig. If there is no odcs configured in
+            reactor config, None is returned.
+        :rtype: :class:`ODCSConfig` or None
+        """
+        if self.odcs:
+            return ODCSConfig(
+                signing_intents=self.odcs['signing_intents'],
+                default_signing_intent=self.odcs['default_signing_intent']
+            )
+
+    def is_default(self):
+        return self.conf == self.DEFAULT_CONFIG
+
+    def _get_value(self, name, fallback=NO_FALLBACK):
+        try:
+            # make a deep copy to prevent plugins from changing the value for other plugins
+            value = deepcopy(self.conf[name])
+            return value
+        except KeyError:
+            if fallback != NO_FALLBACK:
+                return fallback
+            raise
+
+    @property
+    def koji(self):
+        koji_map = self._get_value(ReactorConfigKeys.KOJI_KEY)
+
+        if 'auth' in koji_map:
+            krb_principal = koji_map['auth'].get('krb_principal')
+            krb_keytab = koji_map['auth'].get('krb_keytab_path')
+            if bool(krb_principal) != bool(krb_keytab):
+                raise RuntimeError("specify both koji_principal and koji_keytab or neither")
+
+        return koji_map
+
+    @property
+    def koji_session(self):
+        from atomic_reactor.utils.koji import create_koji_session
+
+        auth_info = {
+            "proxyuser": self.koji['auth'].get('proxyuser'),
+            "ssl_certs_dir": self.koji['auth'].get('ssl_certs_dir'),
+            "krb_principal": self.koji['auth'].get('krb_principal'),
+            "krb_keytab": self.koji['auth'].get('krb_keytab_path')
+        }
+
+        use_fast_upload = self.koji.get('use_fast_upload', True)
+
+        return create_koji_session(self.koji['hub_url'], auth_info, use_fast_upload)
+
+    @property
+    def koji_path_info(self):
+        from koji import PathInfo
+
+        # Make sure koji root_url doesn't end with a slash since the url
+        # is used to construct resource urls (e.g. log links)
+        root_url = self.koji['root_url'].rstrip('/')
+        return PathInfo(topdir=root_url)
+
+    @property
+    def pnc(self):
+        return self._get_value(ReactorConfigKeys.PNC_KEY)
+
+    @property
+    def odcs(self):
+        return self._get_value(ReactorConfigKeys.ODCS_KEY, fallback={})
```

Comment:
It will be confusing if some properties have fallbacks and others don't.
```python
>>> conf = Configuration(workflow, raw_config={})
>>> conf.odcs
{}
>>> conf.pnc
KeyError: 'pnc'  # why does one fail and the other does not, if neither was in the config?
```
Could this be done more consistently?

Reply:
that is based on schema default values

before we had fallback randomly used in plugins, now we have it only based on schema default values, and without forcing it

Reply:
In the schema, these top-level properties have a `default` value:
```
group_manifests
prefer_schema1_digest
skip_koji_check_for_base_image
deep_manifest_list_inspection
fail_on_digest_mismatch
list_rpms_from_scratch
allow_multiple_remote_sources
```
A lot of properties have fallbacks and no default in the schema

Reply:
ok I'll add fallback to everything except of required keys in the schema

Reply:
changed